packages/compiler/lib/lib.cadl

```diff
@@ -1,8 +1,31 @@
 namespace Cadl;
 
+@intrinsic("any")
+model any {}
+
+@intrinsic("object")
+model object {}
+
+@intrinsic("Record")
+model Record<T> {
```

Comment:
I wonder if a record should be a specialized model type with a property that says what its indexer type is (instead of squirreling that info away in a model property).

Reply:
yeah I have been thinking for that array(would help with having `model is string[]`), map, etc that maybe we should have a `ModelType` with a `modelKind` that can be `record`, `array`, `map`, `primitive` and would let us have discriminated type for those known "model".

Reply:
The `indexer` might be the primitive here? A record is a model with an index type of string, an array is a model with an indexer type of integer. And maybe primitives have an indexer type of never?

Reply:
I like that

Reply:
Reserve judgement until you see how many things this breaks 😂 (but if it's good we should do it anyway)

Reply:
I think there is a little issue I just discovered when writting some typescript. In TS you cannot assign an interface to Record because the interface allows additional properties which could then not be matching the record. However if you create the type using `type =` then it is exactly that type.

In the current state of implementation you can asisgn a model to a Record as long as the property types all match. This could lead to this scenario that is not valid

```cadl
model Foo<T is Record<string, string>> {

}

model MyModel {
  prop1: string;
  prop2: string
}

model Bar<T is MyModel> {
  foo: Foo<T>
}


model Test {
  // This would compile but now Foo is receive a number property
  a: Bar<{prop1: string, prop2: string, prop3: int32}>
}
```

Reply:
I feel like this could work with our "plug and chug" templates - instantiating Bar, we check if it extends MyModel, and it does, so all good. Instantiate Foo, we check if it extends Record<string>, but it doesn't, so that check fails. Getting a good error of course depends on showing a stack trace of some sort.

(really want to change is to extends fwiw)

Reply:
the other option is to require having `model MyModel is Record<string>` to say the indexer of MyModel got to keep being a string.

Reply:
I want to amend my previous statement from "could work" to "should work" - is there a reason that it shouldn't? Since we are going with dynamic evaluation of template arguments ala C++ templates, it feels like it should be fine...

Reply:
Sure, feel like this is more of a correctness issue.

Reply:
ok, so I updated the record to not be an intrinsic type but have an `indexer` which can then be used to validate assignment. This works great. Not sure if making the array a model type as well with a number index should be in this PR or a follow up.